src/NuGet.Services.AzureSearch/DownloadTransferrer.cs

```diff
@@ -64,6 +113,143 @@ public DownloadTransferrer(ILogger<DownloadTransferrer> logger)
+            //    Update all new "to" packages
+            //    Update all old "to" packages (in case "to" packages were removed)
+            foreach (var transferChange in transferChanges)
+            {
+                var fromPackage = transferChange.Key;
+                var toPackages = transferChange.Value;
+
+                affectedPackages.Add(fromPackage);
+                affectedPackages.UnionWith(toPackages);
+
+                if (oldOutgoingTransfers.TryGetValue(fromPackage, out var oldToPackages))
+                {
+                    affectedPackages.UnionWith(oldToPackages);
+                }
+            }
+
+            // If a package has download changes and outgoing transfers
+            //    Update "from" package
+            //    Update all "to" packages
+            //
+            // If a package has download changes and incoming transfers
+            //    Update "to" package
+            foreach (var packageId in downloadChanges.Keys)
+            {
+                if (outgoingTransfers.TryGetValue(packageId, out var toPackages))
+                {
+                    affectedPackages.Add(packageId);
+                    affectedPackages.UnionWith(toPackages);
+                }
+
+                if (incomingTransfers.ContainsKey(packageId))
+                {
+                    affectedPackages.Add(packageId);
+                }
+            }
+
+            return affectedPackages;
+        }
+
+        private long TransferPackageDownloads(
+            string packageId,
+            SortedDictionary<string, SortedSet<string>> outgoingTransfers,
+            SortedDictionary<string, SortedSet<string>> incomingTransfers,
+            DownloadData downloads)
+        {
+            var originalDownloads = downloads.GetDownloadCount(packageId);
+            var transferPercentage = _options.Value.Scoring.PopularityTransfer;
+
+            // Calculate packages with outgoing transfers first. These packages transfer a percentage
+            // or their downloads equally to a set of "incoming" packages. Packages with both outgoing
+            // and incoming transfers "reject" the incoming transfers.
+            if (outgoingTransfers.ContainsKey(packageId))
+            {
+                var keepPercentage = 1 - transferPercentage;
+
+                return (long)(originalDownloads * keepPercentage);
+            }
+
+            // Next, calculate packages with incoming transfers. These packages receive downloads
+            // from one or more "outgoing" packages.
+            if (incomingTransfers.TryGetValue(packageId, out var incomingTransferIds))
+            {
+                var result = originalDownloads;
+
+                foreach (var incomingTransferId in incomingTransferIds)
+                {
+                    var incomingDownloads = downloads.GetDownloadCount(incomingTransferId);
+                    var incomingSplit = outgoingTransfers[incomingTransferId].Count;
```

Comment:
Is it possible for `incomingTransferId` to not be present in `outgoingTransfers`?

Reply:
`incomingTransfers` is generated by inversing `outgoingTransfers`, so no, that should never happen.